lib/Conversion/TorchToLinalg/AdaptivePooling.cpp

```diff
@@ -0,0 +1,237 @@
+    2. gather most of the boilerplate out of this op and make it into an
+adaptive pooling helper function.
+    3. figure out what to do with the conflicting decompositions in
+DecomposeComplexOps.cpp
+    4. Implement more efficient passes for when the kernel-size, input spatial
+dims, and output spatial dims are constant.
+*/
+
+namespace {
+class ConvertAtenAdaptiveAvgPool1dOp
+    : public OpConversionPattern<AtenAdaptiveAvgPool1dOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenAdaptiveAvgPool1dOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Location loc = op->getLoc();
+    const TypeConverter *typeconverter = getTypeConverter();
+
+    // input tensor and output shape
+    Value input = adaptor.getSelf();
+    Value outputShape = adaptor.getOutputSize();
+    SmallVector<Value> outShapeVector;
+    getListConstructElements(outputShape, outShapeVector);
+    outShapeVector =
+        getTypeConvertedValues(rewriter, loc, typeconverter, outShapeVector);
+    Value Hin = getDimOp(rewriter, loc, input, 2);
+    Value Hout = outShapeVector[0];
+    Value HoutIndex = castIntToIndex(rewriter, loc, Hout);
+    RankedTensorType InputType = input.getType().cast<RankedTensorType>();
+    RankedTensorType OutputType =
+        typeconverter->convertType(op.getResult().getType())
+            .cast<RankedTensorType>();
+
+    // get rank of input (same as rank of output)
+    int64_t rank = input.getType().cast<RankedTensorType>().getRank();
+    // input operand should be NCH (i.e. rank 3)
+    if (rank != 3) {
+      return rewriter.notifyMatchFailure(op, "only supports input type NCH");
+    }
+
+    // get elementType of input tensor
+    Type elementType = InputType.getElementType();
+
+    // make an iteration space of size Kmax = 1 + ceildiv (Hin - 1) , Hout
+    Type dummyType = rewriter.getI1Type();
+    Value Kiter;
+    if (InputType.isDynamicDim(rank - 1) || OutputType.isDynamicDim(rank - 1)) {
+      Value constantOne =
+          rewriter.create<arith::ConstantOp>(loc, rewriter.getIndexAttr(1));
+      Value HinPlusOne = rewriter.create<arith::SubIOp>(loc, Hin, constantOne);
+      Value KmaxMinusOne =
+          rewriter.create<arith::CeilDivSIOp>(loc, HinPlusOne, HoutIndex);
+      Value Kmax =
+          rewriter.create<arith::AddIOp>(loc, constantOne, KmaxMinusOne);
+      Kiter = rewriter.create<tensor::EmptyOp>(
+          loc, RankedTensorType::get({ShapedType::kDynamic}, dummyType), Kmax);
+    } else {
+      int64_t HinInt = InputType.getShape()[rank - 1];
+      int64_t HoutInt = OutputType.getShape()[rank - 1];
+      int64_t correction = ((HinInt - 1) % HoutInt == 0) ? 0 : 1;
+      int64_t Kmax = 1 + (HinInt - 1) / HoutInt + correction;
+      Kiter = rewriter.create<tensor::EmptyOp>(
+          loc, RankedTensorType::get({Kmax}, dummyType), ValueRange{});
+    }
```

Comment:
No need to have special handling for static shapes. If the sizes are static, a backend compiler should be able to easily optimize away the ops generated in the dynamic case.

Reply:
Ah! This is very helpful, thanks.